### DISCUSSION TITLE
Wrong tokens probabilities

### DISCUSSION THREAD

**Discussion by Rtut654** [openai/whisper#1591] [Q&A]
Hello. I used whisper version from this [PR](https://github.com/openai/whisper/pull/1119), only difference - I changed weights to "small". I'm facing the following issue:
I made distorted speech example https://drive.google.com/file/d/12zGWllJg6edftcnwuHX_ZHMuwk7PlVjg/view?usp=sharing .
If I don't set the language of decoding i.e. `options = whisper.DecodingOptions() `, the output is correct in terms of locating mispronounce (I can read this Ukrainian) though it translates it to random language.
![Screenshot from 2023-08-09 10-58-49](https://github.com/openai/whisper/assets/54935496/1eabe017-4d7c-400e-ad11-107687d4c536)
But if I set 'en' for decoding `options = whisper.DecodingOptions(language="en")` the picture and probabilities are wrong.
![Screenshot from 2023-08-09 10-58-53](https://github.com/openai/whisper/assets/54935496/d5734b89-ac26-42ad-b310-632cb0898638) .
I checked a few similar audios and result is almost same. Probs are high for each word except the last one.

**Comment by phineas-pta** [comment]
this feature isnt official yet, u have to ask its author

**Reply by Rtut654** [reply]
It is not about this feature, the code in PR is fairly simple with only visualization added. The question is about whisper's strange behavior

**Reply by phineas-pta** [reply]
did u use original whisper ? it's same results ?

**Reply by Rtut654** [reply]
Yes, the result is the same, I just added a few lines to get `token_probs `:
```
 def update(
        self, tokens: Tensor, logits: Tensor, sum_logprobs: Tensor, token_probs: Tensor
    ) -> Tuple[Tensor, bool]:
        if self.temperature == 0:
            next_tokens = logits.argmax(dim=-1)
        else:
            next_tokens = Categorical(logits=logits / self.temperature).sample()

        logprobs = F.log_softmax(logits.float(), dim=-1)
        probs = torch.exp(logprobs)
        
        current_logprobs = logprobs[torch.arange(logprobs.shape[0]), next_tokens]

        next_tokens[tokens[:, -1] == self.eot] = self.eot
        tokens = torch.cat([tokens, next_tokens[:, None]], dim=-1)

        current_token_probs = probs[torch.arange(probs.shape[0]), next_tokens]
        token_probs = torch.cat([token_probs, current_token_probs[:, None]], dim=-1)
        
        completed = (tokens[:, -1] == self.eot).all()
        return tokens, completed, token_probs
```

**Reply by phineas-pta** [reply]
wait a sec, u set `language="en"` should have used `task="translate"`

**Reply by Rtut654** [reply]
Haven't got why, I want to transcribe audio in english to english text. Isn't it `'transcribe'` task?

**Reply by phineas-pta** [reply]
nvm, i thought the audio is ukrainian not english

the quality is so bad that whisper misdetect the language, did u try better quality audio that whisper detect language correctly ?
